src/Crypto/Secp256k1.hs

```diff
@@ -743,23 +745,31 @@ keyPairPubKeyXOTweakAdd KeyPair{..} Tweak{..} = unsafePerformIO . evalContT $ do
             else free keyPairOut $> Nothing
 
 
--- | Compute a schnorr signature using a 'KeyPair'. The @ByteString@ must be 32 bytes long to get a @Just@ out of this
--- function
-schnorrSign :: KeyPair -> ByteString -> IO (Maybe SchnorrSignature)
-schnorrSign KeyPair{..} bs
-    | BS.length bs /= 32 = pure Nothing
-    | otherwise = evalContT $ do
+-- | Compute a schnorr signature using a 'KeyPair'. The @ByteString@ must be 32 bytes long to get 
+-- a @Just@ out of this function. Optionally takes a 'StdGen' for deterministic signing.
+schnorrSign :: Maybe StdGen -> KeyPair -> ByteString -> Maybe SchnorrSignature
+schnorrSign mGen KeyPair{..} bs
+    | BS.length bs /= 32 = Nothing
+    | otherwise = unsafePerformIO . evalContT $ do
         (msgHashPtr, _) <- ContT (unsafeUseByteString bs)
         keyPairPtr <- ContT (withForeignPtr keyPairFPtr)
         lift $ do
             sigBuf <- mallocBytes 64
-            gen <- newStdGen
-            let randomBytes = BS.pack $ Prelude.take 32 $ randoms gen
-            BS.useAsCString randomBytes $ \randomPtr -> do
-                ret <- Prim.schnorrsigSign ctx sigBuf msgHashPtr keyPairPtr (castPtr randomPtr)
-                if isSuccess ret
-                    then Just . SchnorrSignature <$> newForeignPtr finalizerFree sigBuf
-                    else free sigBuf $> Nothing
+            let randomPtr = case mGen of
+                    Just gen -> castPtr . unsafePerformIO . BS.useAsCString (BS.pack $ Prelude.take 32 $ randoms gen) $ return
```

Comment:
don't use unsafePerformIO here. The toplevel one is OK, but each invocation of it is ok. On this line, just fmap the castPtr and then in the Nothing branch use pure/return.

You also cannot apply this to useAsCString to return because otherwise it leaks the CString outside of the safe scope: see [this](https://hackage.haskell.org/package/bytestring-0.12.1.0/docs/Data-ByteString.html#v:useAsCString)

Reply:
done

Reply:
You still need to fix the `useAsCString` being applied to return. it leaks the value which you cannot do.

Reply:
like this?